### PR TITLE
Add "data" event

### PR DESCRIPTION
**Pull request by lucaswoj**
This PR consolidates various undocumented data lifecycle events into the new `data` and `dataloading` events. 

 - [x] replace `source.change` with `data`
 - [x] augment `source.load` with `data`
 - [x] replace `style.change` with `data` 
 - [x] augment `style.load` with `data`
 - [x] replace `tile.add` with `dataloading`
 - [x] replace `tile.load` with `data`
 - [x] replace `tile.remove`with `data`
 - [x] replace `source.add` with `data`
 - [x] replace `source.remove` with `data`
 - [x] replace `layer.add` with `data`
 - [x] replace `layer.remove` with `data`
 - [x] document `data` event
 - [x] document `dataloading` event
 - [x] merge `geoJSON`, `tileJSON`, `image`, and `video` types into a `source` type
 - [x] merge `sprite` into the `style` type

Having an event per resource type per possible mutation produced a complex taxonomy of events. Understanding, documenting, and using these events was difficult. Most users did not require the level of granularity exposed by this taxonomy. Some use cases required listening to many different events. 

The unified `data` event is easy to understand, document, and use. Users who require more granularity may inspect the `MapDataEvent` object. 

> A `MapDataEvent` object is attached to the [`Map#data`](#Map.event:data) event.
> The `Map#data` event is fired when any map data loads or changes. The types
> of map data are:
> 
> - `'geoJSON'`: [GeoJSON](http://geojson.org/) data associated with a `geojson` source.
> - `'tileJSON'`: [TileJSON](https://github.com/mapbox/tilejson-spec) metadata associated with a `vector` or `raster` source.
> - `'style'`: The [style](https://www.mapbox.com/mapbox-gl-style-spec/) used by the map
> - `'sprite'`: The icons and patterns used by the style
> - `'image'`: A `image` source's image and coordinates
> - `'video'`: A `video` source's video and coordinates
> - `'tile'`: A vector or raster tile

cc @jfirebaugh @lbud @mollymerp @mapsam @mourner 

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] document any changes to public APIs
 - [x] post benchmark scores
 - [x] manually test the debug page

### CODE REVIEW

**Comment by lucaswoj** [comment]
# Benchmarks

## map-load

**master 24f2135:** 119 ms
**datastart 1ef4213:** 105 ms

## style-load

**master 24f2135:** 109 ms
**datastart 1ef4213:** 113 ms

## buffer

**master 24f2135:** 961 ms
**datastart 1ef4213:** 941 ms

## fps

**master 24f2135:** 60 fps
**datastart 1ef4213:** 60 fps

## frame-duration

**master 24f2135:** 6.9 ms, 0% > 16ms
**datastart 1ef4213:** 7.2 ms, 1% > 16ms

## query-point

**master 24f2135:** 1.04 ms
**datastart 1ef4213:** 1.12 ms

## query-box

**master 24f2135:** 85.34 ms
**datastart 1ef4213:** 85.61 ms

## geojson-setdata-small

**master 24f2135:** 13 ms
**datastart 1ef4213:** 14 ms

## geojson-setdata-large

**master 24f2135:** 125 ms
**datastart 1ef4213:** 121 ms

**Comment by jfirebaugh** [comment]
Can you post a full rundown of the new taxonomy -- what are the valid combinations of event types and `dataType` values, and when are they emitted?

**Comment by lucaswoj** [comment]
Below is the type specification for the `data` and `dataloading` events:

```js
interface DataEvent {
    type: 'dataloading' | 'data';
    dataType: (
        'geoJSON' |
        'tileJSON' |
        'style' |
        'sprite' |
        'image' |
        'video' |
        'tile'
    );
}
```

The `Map#data` event is fired when any map data loads or changes.

The `Map#dataloading` event is fired when any map data begins loading or changing asynchronously, to be followed by a `data` event when the operation is complete.

**Comment by jfirebaugh** [comment]
But there's also `source.load` and `style.load` still?

**Comment by jfirebaugh** [comment]
What's the rationale for having separate `geoJSON`, `tileJSON`, `image`, and `video` values for `dataType`, rather than a single `source` value? It seems like a `source` value would simplify the taxonomy and keep it consistent with the domain objects used in rest of the API and in the style specification.

File: js/source/image_source.js [hunk 1]
```diff
@@ -53,6 +53,7 @@ function ImageSource(id, options, dispatcher) {
 
         this.image = image;
         this._loaded = true;
+        this.fire('data', {type: 'image'});
```
Comment (jfirebaugh) [inline js/source/image_source.js]: `type` ⇢ `dataType`

File: js/source/raster_tile_source.js [hunk 2]
```diff
@@ -17,6 +17,7 @@ function RasterTileSource(id, options, dispatcher) {
             return this.fire('error', err);
         }
         util.extend(this, tileJSON);
+        this.fire('data', {type: 'tileJSON'});
```
Comment (jfirebaugh) [inline js/source/raster_tile_source.js]: `type`  ⇢ `dataType`

**Comment by lucaswoj** [comment]
> But there's also `source.load` and `style.load` still?

The plan is to have `data` subsume those events in a follow-up PR. See the plan in https://github.com/mapbox/mapbox-gl-js/issues/1715#issuecomment-249293404

> What's the rationale for having separate `geoJSON`, `tileJSON`, `image`, and `video` values for `dataType`, rather than a single `source` value? It seems like a `source` value would simplify the taxonomy and keep it consistent with the domain objects used in rest of the API and in the style specification.

The rationale was to have a `dataType` per specific network resource type. Merging `geoJSON`, `tileJSON`, `image`, and `video` into `source` is a good idea. I'll push a commit shortly.

File: bench/lib/set_data_perf.js [hunk 3]
```diff
@@ -7,7 +7,9 @@ module.exports = function(source, numCalls, geojson, cb) {
     var startTime = null;
     var times = [];
 
-    source.on('tile.load', function tileCounter() {
+    source.on('data', function tileCounter(event) {
+        if (event.dataType !== 'tile') return;
```
Comment (jfirebaugh) [inline bench/lib/set_data_perf.js]: Does this event binding need to distinguish between "tile loading" and "tile unloading"?

File: js/style/style.js [hunk 4]
```diff
@@ -22,8 +22,9 @@ var getWorkerPool = require('../global_worker_pool');
 
 module.exports = Style;
 
-function Style(stylesheet, animationLoop, options) {
-    this.animationLoop = animationLoop || new AnimationLoop();
+function Style(stylesheet, map, options) {
+    this.map = map;
```
Comment (jfirebaugh) [inline js/style/style.js]: I've been trying to reduce or eliminate the dependency of other class on `Map`. The dependency here is required by `Source#on{Add,Remove}(map)` -- we should work towards eliminating those two methods.

File: js/ui/map.js [hunk 5]
```diff
@@ -197,23 +197,11 @@ var Map = module.exports = function(options) {
         this.style.update(this._classes, {transition: false});
     });
 
-    this.on('style.change', function() {
+    this.on('data', function() {
         this._update(true);
```
Comment (jfirebaugh) [inline js/ui/map.js]: Not every `data` event should trigger `_update(true)` -- that triggers updating and recalculating, which is necessary only in certain cases.

File: bench/lib/set_data_perf.js (hunk 3 above)
Comment (lucaswoj) [inline bench/lib/set_data_perf.js]: I think so. I will refactor to make this benchmark utility more robust.

File: js/style/style.js (hunk 4 above)
Comment (lucaswoj) [inline js/style/style.js]: Removing the dependency was my first instinct as well. Unfortunately it wasn't straightforward to do so and will require a standalone PR. 

Even though the dependency is undesirable, it is a real dependency and should be passed explicitly and simply. As our code is architected, `Source`s need to know about the state of `Camera` (which is a mixin on `Map`). I see #2741 as the proper long-term fix.

File: js/ui/map.js (hunk 5 above)
Comment (lucaswoj) [inline js/ui/map.js]: Good catch. Fortunately it looks like filtering on `event.dataType === 'style'` is equivalent to listening to the `style.change` event.

File: js/source/raster_tile_source.js (hunk 2 above)
Comment (lucaswoj) [inline js/source/raster_tile_source.js]: Fixed

File: js/source/image_source.js (hunk 1 above)
Comment (lucaswoj) [inline js/source/image_source.js]: Fixed

**Comment by lucaswoj** [comment]
All feedback addressed, ready for a second pass of  👀 @jfirebaugh

File: js/ui/control/attribution.js [hunk 6]
```diff
@@ -54,9 +54,7 @@ Attribution.prototype = util.inherit(Control, {
             container = this._container = DOM.create('div', className, map.getContainer());
 
         this._update();
-        map.on('source.load', this._update.bind(this));
-        map.on('source.change', this._update.bind(this));
-        map.on('source.remove', this._update.bind(this));
+        map.on('data', this._update.bind(this));
```
Comment (jfirebaugh) [inline js/ui/control/attribution.js]: Add a `dataType === 'source'` guard in `_update` so that it doesn't mutate the DOM more frequently than it did before.

Comment (lucaswoj) [inline js/ui/control/attribution.js]: Done.

**Comment by lucaswoj** [comment]
Thank you for the review @jfirebaugh 🙇

**Comment by strech345** [comment]
is there a 'unload' function?
i need the event which tile is unload.

**Comment by lucaswoj** [comment]
@strech345 Not right now. Defining what it means for a tile to become "unloaded" isn't easy. Is a tile unloaded

 - When it is no longer at the desired zoom level for the current viewport but still rendered?
 - When it is entirely outside of the current viewport?
 - When it is removed from the cache? How big is the cache? Which cache?
 - When it is reloaded?

**Comment by strech345** [comment]
@lucaswoj it the same question like for the loading. So i think it should be a equivalent to the dataloading
